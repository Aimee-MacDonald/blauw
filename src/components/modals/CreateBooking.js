import React, {useContext} from 'react'
import styled from 'styled-components'
import {connect} from 'react-redux'

import {ServerConnectionContext} from '../../util/ServerConnection'

import {setModal} from '../../state_management/actions/modal'
import {createBooking} from '../../state_management/actions/bookings'

const CreateBooking = props => {
  const connection = useContext(ServerConnectionContext)

  const saveBooking = e => {
    e.preventDefault()

    const bookingDetails = {
      name: e.target.name.value,
      date: parseInt(e.target.date.value),
      room: parseInt(e.target.room.value),
      nights: parseInt(e.target.nights.value)
    }

    const createBookingAction = createBooking(bookingDetails)

    props.dispatch(createBookingAction)
    connection.dispatch({reducer: 'bookings', action: createBookingAction})
    props.dispatch(setModal())
  }

  return(
    <StyledCreateBooking onSubmit={saveBooking}>
      <h1>Create Booking</h1>

      <label htmlFor='name'>Name:</label>
      <input id='name' />

      <label htmlFor='date'>Date:</label>
      <input id='date' />

      <label htmlFor='room'>Room:</label>
      <input id='room' />

      <label htmlFor='nights'>Nights:</label>
      <input id='nights' />
      
      <button type='reset' onClick={() => props.dispatch(setModal())}>Cancel</button>
      <button type='submit'>Save</button>
    </StyledCreateBooking>
  )
}

const StyledCreateBooking = styled.form`
  background-color: lime;
`

export default connect()(CreateBooking)