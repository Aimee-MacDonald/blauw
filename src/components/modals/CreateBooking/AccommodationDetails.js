import React from 'react'
import styled from 'styled-components'
import {connect} from 'react-redux'

import {getRoomDetails, getRoomOptions} from '../../../state_management/selectors/rooms'

export const AccommodationDetails = props => {
  const setGuestName = (roomIndex, guestIndex, guestName) => {
    props.editGuestDetails(roomIndex, guestIndex, {
      ...props.rooms[roomIndex].guests[guestIndex],
      name: guestName
    })
  }


  const setGuestArrival = (roomIndex, guestIndex, guestArrival) => {
    props.editGuestDetails(roomIndex, guestIndex, {
      ...props.rooms[roomIndex].guests[guestIndex],
      arrival: guestArrival
    })
  }

  const setGuestDeparture = (roomIndex, guestIndex, guestDeparture) => {
    props.editGuestDetails(roomIndex, guestIndex, {
      ...props.rooms[roomIndex].guests[guestIndex],
      departure: guestDeparture
    })
  }

  return(
    <StyledAccommodationDetails>
      <p>Room</p>
      <p>Guests</p>
      <p>Arrival</p>
      <p>Departure</p>
      <p>Nights</p>
      <p>Price</p>

      {props.rooms && props.rooms.map(room => (
        <StyledRoomDetails key={room.index}>
          <select onChange={e => props.selectRoomOption(e.target.value)}>
            <option>Select Room</option>
            {getRoomOptions(props.stateRooms).map(roomOption => (
              <option key={roomOption._id} value={roomOption._id}>{roomOption.name}</option>
            ))}
          </select>

          {room.guests.map((guest, index) => (
            <StyledGuestDetails key={`${guest}_${index}`}>
              <input
                value={guest.name}
                onChange={e => setGuestName(room.index, index, e.target.value)}
              />

              <input
                type='date'
                value={guest.arrival}
                onChange={e => setGuestArrival(room.index, index, e.target.value)}
              />

              <input
                type='date'
                value={guest.departure}
                onChange={e => setGuestDeparture(room.index, index, e.target.value)}
              />

              <p>{guest.nights}</p>
              <p>{`R ${guest.price}`}</p>
            </StyledGuestDetails>
          ))}
        </StyledRoomDetails>
      ))}

      <div id='addButtons'>
        <button type='button' onClick={props.addRoom}>Add Room</button>
        {props.selectedRoom &&
          !!props.rooms[props.selectedRoom - 1].roomId &&
          props.rooms[props.selectedRoom - 1].guests.length < getRoomDetails(props.rooms[props.selectedRoom - 1].roomId, props.stateRooms).maxPax &&
          <button type='button' onClick={props.addGuest}>Add Guest</button>
        }
      </div>

      <p id='total'>Total: R</p>
    </StyledAccommodationDetails>
  )
}

const StyledAccommodationDetails = styled.div`
  grid-column-start: 1;
  grid-column-end: 5;
  display: grid;
  grid-template-columns: repeat(6, 1fr);
  width 64rem;
  text-align: center;

  #addButtons{
    grid-column-start: 1;
    grid-column-end: 3;
    display: flex;
    justify-content: space-evenly;
  }

  #total{
    grid-column-start: 6;
  }
`

const StyledRoomDetails = styled.div`
  grid-column-start: 1;
  grid-column-end: 7;
  display: grid;
  grid-template-columns: repeat(6, 1fr);
`

const StyledGuestDetails = styled.div`
  grid-column-start: 2;
  grid-column-end: 7;
  display: grid;
  grid-template-columns: repeat(5, 1fr);
`

const mapStateToProps = ({rooms}) => ({stateRooms: rooms})

export default connect(mapStateToProps)(AccommodationDetails)