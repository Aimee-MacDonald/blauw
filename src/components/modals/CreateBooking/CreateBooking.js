import React from 'react'
import styled from 'styled-components'

import AccommodationDetails from './AccommodationDetails'

export const CreateBooking = () => (
  <StyledCreateBooking>
    <h1>Create Booking</h1>

    <div id='labeled_booking_name'>
      <label htmlFor='booking_name'>Booking Name</label>
      <input id='booking_name'/>
    </div>

    <AccommodationDetails/>

    <div id='labeled_notes'>
      <label htmlFor='notes'>Notes</label>
      <textarea/>
    </div>

    <div id='formControls'>
      <button>Save</button>
      <button>Cancel</button>
    </div>
  </StyledCreateBooking>
)

const StyledCreateBooking = styled.form`
  display: grid;
  grid-template-columns: repeat(4, 1fr);

  h1{
    grid-column-start: 1;
    grid-column-end: 3;
  }

  #labeled_booking_name{
    grid-column-start: 1;
    grid-column-end: 5;
    display: flex;
  }

  #labeled_notes{
    grid-column-start: 1;
    grid-column-end: 5;
    display: flex;
    flex-flow: nowrap column;

    textarea{
      width: 100%;
    }
  }

  #formControls{
    grid-column-start: 2;
    grid-column-end: 4;
    display: flex;
    flex-flow: nowrap row;
    justify-content: space-evenly;
  }
`

export default CreateBooking