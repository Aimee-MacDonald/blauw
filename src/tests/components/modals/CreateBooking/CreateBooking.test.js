import React from 'react'
import {shallow} from 'enzyme'

import {CreateBooking} from '../../../../components/modals/CreateBooking/CreateBooking'

test(`Render default 'Create Booking' modal`, () => {
  const wrapper = shallow(<CreateBooking/>)
  expect(wrapper.getElement()).toMatchSnapshot()
})