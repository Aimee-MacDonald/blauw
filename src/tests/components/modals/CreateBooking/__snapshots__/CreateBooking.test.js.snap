// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render default 'Create Booking' modal 1`] = `
<styled.form>
  <h1>
    Create Booking
  </h1>
  <div
    id="labeled_booking_name"
  >
    <label
      htmlFor="booking_name"
    >
      Booking Name
    </label>
    <input
      id="booking_name"
    />
  </div>
  <Memo(Connect(AccommodationDetails)) />
  <div
    id="labeled_notes"
  >
    <label
      htmlFor="notes"
    >
      Notes
    </label>
    <textarea />
  </div>
  <div
    id="formControls"
  >
    <button>
      Save
    </button>
    <button>
      Cancel
    </button>
  </div>
</styled.form>
`;
